### PR TITLE
sql: native evaluation support for NotExpr

### LINKED ISSUE #70713: colexec: implement vectorized version of NotExpr
Currently, `tree.NotExpr` is not implemented natively in the vectorized engine, so we have to fallback to the older row-by-row engine to evaluate it. We should, instead, vectorize `NotExpr`.

I think the implementation will be quite similar to how `tree.IsNullExpr` is implemented, and I think we will want to implement two versions of `NotExpr` operator:
- one for projections, which populates a `coldata.Vec` with values equal to evaluating `NOT` predicate on each row in another `coldata.Vec` (while paying attention to nulls in the input vector and the selection vector on the input batch)
- one for selections, which updates the selection vector on the batch to include only rows for which `NOT` predicate evaluated to `true`.

Some code pointers:
- the planning is done in `colbuilder/execplan.go` in `planProjectionOperators` and `planSelectionOperators`
- `tree.NotExpr` lives in `sem/tree/expr.go`
- the vectorized implementation of `tree.IsNullExpr` lives in `colexec/is_null_ops.eg.go` (which is automatically generated based on `colexec/is_nulls_ops_tmpl.go` and `colexec/execgen/cmd/execgen/is_nulls_ops_gen.go`).

Please let me know if more guidance is needed.

### PR DESCRIPTION
**Pull request by RajivTS**
The commit includes the following changes to provide native evaluation support for tree.NotExpr:
1. Defined new operators NotExprProjOp for projection and NotExprSelOp for selection when evaluating the result of a NotExpr
2. Defined NotNullProjOp for projection when the underlying type is non-bool and contains only Nulls.
3. Defined the Next method for both the projection and selection operators.
4. Added test cases for testing the functionality of NotExprProjOp, NotNullProjOp and NotExprSelOp operators.

Fixes: #70713

Release note (performance improvement): queries using `NOT expr` syntax can now be evaluated faster in some cases.

### CODE REVIEW

**Comment by RajivTS** [comment]
Hi @yuzefovich 

I closed the previous PR owning to commit squash problems and created this new PR with a single commit and the proper commit message and format. However, the build is failing due to some errors while executing benchmark tests. I am unable to understand the cause of these failures and their relation to the changes made in this PR. Any help would be greatly appreciated!

**Review by yuzefovich — COMMENTED** [review]
Nice work! I have a few more nits as well as the suggestion on how to fix the build failures.

Also, in the PR description if you use the following syntax:
```
Fixes: #70713.
```
then that issue will be closed automatically when the PR merges.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-:-MtjKw6X5jrJNrShefsQ:b8z3ri1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RajivTS)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtjJv9X6HAlYM3pjDod):*
nit: we tend to use lower case.
___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtjJzFM9MIb6r64TfvT):*
nit: wrap lines at 80 characters.
___
*[-- commits, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtjK33s8pIkYZ5_xWL6):*
nit: this release note is a user-facing piece of documentation (this is what will be included into the Release notes web-page that accompanies the first release that includes this fix), so it should be written as if you were talking to a user that doesn't know much about the internals. So I would rephrase it with something like:
```
Release note (performance improvement): queries using `NOT expr` syntax can now
be evaluated faster in some cases.
```
___
*[pkg/sql/colexec/not_expr_ops.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtjJ3SS7RXi2m9BtEjf:-MtjJ3SS7RXi2m9BtEjg:b1asw68) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0783bb396655e0c8e49e4354d3f9fa268460cb1/pkg/sql/colexec/not_expr_ops.go#L21)):*
> ```Go
> )
> 
> type notExprProjBase struct {
> ```

nit: no need for "base" structs here since they are used only in one place - we should just inline the contents directly into the operator struct.
___
*[pkg/sql/colexec/colbuilder/execplan.go, line 1992 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtjLk1sATZViz93mQxp:-MtjLk1sATZViz93mQxq:bcjszxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0783bb396655e0c8e49e4354d3f9fa268460cb1/pkg/sql/colexec/colbuilder/execplan.go#L1992)):*
> ```Go
> 	case *tree.NotExpr:
> 		op, resultIdx, typs, err = planProjectionOperators(
> 			ctx, evalCtx, expr, columnTypes, input, acc, factory, releasables,
> ```

I think this should be `t.Expr` and not `expr`, otherwise we have an infinite recursion:
```
goroutine 113455 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0xc001008ef0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x12, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1945 +0x2ec5 fp=0xc026882410 sp=0xc026882408 pc=0x2c3e3c5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1991 +0x6c5 fp=0xc026882818 sp=0xc026882410 pc=0x2c3bbc5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1991 +0x6c5 fp=0xc026882c20 sp=0xc026882818 pc=0x2c3bbc5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1991 +0x6c5 fp=0xc026883028 sp=0xc026882c20 pc=0x2c3bbc5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1991 +0x6c5 fp=0xc026883430 sp=0xc026883028 pc=0x2c3bbc5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1991 +0x6c5 fp=0xc026883838 sp=0xc026883430 pc=0x2c3bbc5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1991 +0x6c5 fp=0xc026883c40 sp=0xc026883838 pc=0x2c3bbc5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder/execplan.go:1991 +0x6c5 fp=0xc026884048 sp=0xc026883c40 pc=0x2c3bbc5
github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.planProjectionOperators({0x4d04010, 0xc000882680}, 0x0, {0x4d365e0, 0xc003ff5080}, {0xc004257e60, 0x0, 0x12}, {0x4d1d488, 0xc00127e780}, ...)
...
```

**Comment by RajivTS** [comment]
Hi @yuzefovich 

Thank you for the pointers and for being patient with me!

The current version of the PR is still failing due certain unit test failures pertaining to `eval`. Based on my investigation, the NOT NULL variant of the boolean eval test case has only NULL as input and when that is being cast as a slice of booleans by `inputVec.Bool()`, we get a runtime error. Not sure what can be done here except pulling the cast inside for each data item (i.e. the previous inefficient version). Any inputs would be greatly welcome.

image

**Review by yuzefovich — COMMENTED** [review]
No worries, thanks for working through all the failures!

I think we should be able to determine this case at the planning time, and then use new separate operators for it, `notNullOp`. For projection it would simply set the NULLs on all rows, for selection it would select no rows. Take a look at `BoolOrUnknownToSelOp` for inspiration. The way we determine this case at the planning time is by examining the type of the input, so something like the following diff:
```
--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -1978,6 +1978,15 @@ func planProjectionOperators(
                        ctx, evalCtx, t.Operator, t.ResolvedType(), t.TypedLeft(), t.TypedRight(),
                        columnTypes, input, acc, factory, nil /* binFn */, t, releasables,
                )
+       case *tree.NotExpr:
+               switch t.TypedInnerExpr().ResolvedType().Family() {
+               case types.BoolFamily:
+                       // use regular notExprOp
+               case types.UnknownFamily:
+                      // use special notNullProjOp
+               default:
+                       return errors.AssertionFailedf("unexpected type for NOT expr")
+               }
```
For selections we can use `NewZeroOp` directly, but for projections a new separate `notNullProjOp` will need to be written.

Hope this makes sense, let me know if you have any questions.

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-:-Mtokwe72qSYloDTDuzQ:byopwag)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RajivTS)

**Comment by RajivTS** [comment]
@yuzefovich I think I have made the changes that you suggested. Please let me know if there are additional tweaks I need to make. The build is still failing due to git being unable to acquire a lock. Is this due to the changes I have made?

**Review by yuzefovich — COMMENTED** [review]
Thanks! I have some more nits about how to clean this up, but I think it's essentially ready to go!

> The build is still failing due to git being unable to acquire a lock. Is this due to the changes I have made?

I think we had some issues with the CI, it should be good now.

> Reviewed 2 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-:-MtyUOq-7wW9DpwyOS1i:b-ccsqa8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RajivTS)

___
*[pkg/sql/colexec/not_null_ops.go, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtyTa-J0kz_D2UkV5vb:-MtyTa-K0VE0UstGvzjR:b-nlu78u) ([raw file](https://github.com/cockroachdb/cockroach/blob/88360fb015720ab5db7b955077e506d419c04db8/pkg/sql/colexec/not_null_ops.go#L1)):*
> ```Go
> // Copyright 2022 The Cockroach Authors.
> ```

nit: I think it'd be cleaner if we merged this file into `not_expr_ops` because `notNull` is just a special case of `NOT expr`.
___
*[pkg/sql/colexec/not_null_ops.go, line 21 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtyTXDEEBFddZnooQpH:-MtyTXDEEBFddZnooQpI:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/88360fb015720ab5db7b955077e506d419c04db8/pkg/sql/colexec/not_null_ops.go#L21)):*
> ```Go
> )
> 
> // notNullProjOp is an Operator that projects NULL into outputIdx Vec
> ```

nit: missing period.
___
*[pkg/sql/colexec/not_null_ops.go, line 51 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtyTo4X3c5TqIeQ0ZS_:-MtyTo4X3c5TqIeQ0ZSa:b-m9vb4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/88360fb015720ab5db7b955077e506d419c04db8/pkg/sql/colexec/not_null_ops.go#L51)):*
> ```Go
> 	}
> 	outputNulls := batch.ColVec(o.outputIdx).Nulls()
> 	if sel := batch.Selection(); sel != nil {
> ```

We can simplify this logic to just use `outputNulls.SetNulls()`, regardless of the selection vector.
___
*[pkg/sql/colexec/colbuilder/execplan.go, line 1854 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtyU33TEabbTsKkqQfZ:-MtyU33TEabbTsKkqQf_:b-xtxgl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/88360fb015720ab5db7b955077e506d419c04db8/pkg/sql/colexec/colbuilder/execplan.go#L1854)):*
> ```Go
> 		switch t.TypedInnerExpr().ResolvedType().Family() {
> 		case types.BoolFamily:
> 			op = colexec.NewNotExprSelOp(op, resultIdx)
> ```

nit: I think it'd be cleaner if we pushed this `switch` into `NewNotExprSelOp`.
___
*[pkg/sql/colexec/colbuilder/execplan.go, line 2005 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtyU9BE4t8vQdywXfsp:-MtyU9BE4t8vQdywXfsq:b-9cy9uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/88360fb015720ab5db7b955077e506d419c04db8/pkg/sql/colexec/colbuilder/execplan.go#L2005)):*
> ```Go
> 		}
> 		outputIdx := len(typs)
> 		switch t.TypedInnerExpr().ResolvedType().Family() {
> ```

nit: ditto for pushing the switch into the constructor.
___
*[pkg/sql/colexec/not_null_ops_test.go, line 41 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-MtyUHVxDpHFO-GJBHIC:-MtyUHVy6GEpENm0Oywd:b-g2uyg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/88360fb015720ab5db7b955077e506d419c04db8/pkg/sql/colexec/not_null_ops_test.go#L41)):*
> ```Go
> 	}{
> 		{
> 			desc:         "SELECT NOT c FROM t -- NOT expr with only NULL",
> ```

nit: can also merge this into the other test for regular `NOT expr`.

**Comment by RajivTS** [comment]
@yuzefovich I have incorporated the suggested changes. Hopefully this commit will do the job :)

**Review by yuzefovich — COMMENTED** [review]
Looks great! Thanks for your work on this! (I made some minor changes to address some of the minor nits like missing periods.)

bors r=yuzefovich

> Reviewed 6 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75058#-:-MuDw472CxwUkdHY5LmK:bhzrx7i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

**Comment by yuzefovich** [comment]
Hm, looks like our CI is not happy for some reason. Let's try again.

bors r=yuzefovich
